libcst/metadata/tests/test_metadata_wrapper.py

```diff
@@ -41,3 +43,34 @@ def test_hash_by_identity(self) -> None:
         self.assertNotEqual(hash(mw1), hash(mw2))
         self.assertNotEqual(hash(mw1), hash(mw3))
         self.assertNotEqual(hash(mw2), hash(mw3))
+
+    def test_metadata_cache(self) -> None:
+        class DummyMetadataProvider(BatchableMetadataProvider[None]):
+            is_cache_required = True
+
+        m = cst.parse_module("pass")
+        mw = MetadataWrapper(m)
+        with self.assertRaisesRegex(
+            Exception, "Cache is required for initializing DummyMetadataProvider."
+        ):
+            mw.resolve(DummyMetadataProvider)
+
+        class SimpleCacheMetadataProvider(BatchableMetadataProvider[object]):
+            is_cache_required = True
+
+            def __init__(self, cache: object) -> None:
+                super().__init__(cache)
+                self.cache = cache
+
+            def visit_Pass(self, node: cst.Pass) -> Optional[bool]:
+                self.set_metadata(node, self.cache)
+
+        cached_data = object()
+        # pyre-fixme[6]: Expected `Mapping[Type[BaseMetadataProvider[object]],
+        #  object]` for 2nd param but got `Dict[Type[SimpleCacheMetadataProvider],
+        #  object]`.
+        mw = MetadataWrapper(m, cache={SimpleCacheMetadataProvider: cached_data})
```

Comment:
Not sure how to correctly type this. CC @bgw

Reply:
This looks like a covariance issue. Dict/Mapping are invariant in keys. Try a `Typevar` with `covariant=True` and see if that doesn't fix this: https://mypy.readthedocs.io/en/latest/generics.html#variance-of-generic-types

Reply:
The generic type used by BaseMetadataProvider is already defined as covariant. It's not clear to me how to actually fix this.
```
_ProvidedMetadataT = TypeVar("_ProvidedMetadataT", covariant=True)

class BaseMetadataProvider(MetadataDependent, Generic[_ProvidedMetadataT]):
    ...

class SimpleCacheMetadataProvider(BatchableMetadataProvider[object]):
    ...
```

Reply:
Perhaps we can't use a Mapping here? What if we introduce a `MetadataCache` class that holds all of the per-file provider-to-data mappings? We need some other layer that can calculate cache for a list of files, and then grab the cache per-file for transform/metadata instantiation anyway. So if we sidestep the Mapping covariance issue with our own wrapper it might also help build the next layer?

Reply:
We discussed about building a full repo manager to handle generating the per-file provider-to-data given a list. It will come in the next few PRs.

Reply:
Ah, then I don't think its worth solving this issue for now, pyre-ignore seems fine.

Reply:
We discussed about building a full repo manager to handle generating the per-file provider-to-data given a list. It will come in the next few PRs.